### DISCUSSION TITLE
Inherit documentation on implementing types of interface

### DISCUSSION THREAD

**Discussion by coleturner** [graphql-java/graphql-java#3249] [Ideas]
When designing a schema using interfaces, it's possible to find that documenting across the implementing types is a lot of work. Here's an example:

```graphql
interface Animal {
   "The scientific classification for this Animal"
   breed: String
}

type Dog implements Animal {
   breed: String
   knownTricks: [Trick]
}

type Cat implements Animal {
   breed: String
   zoomieSchedule: [DateTimeRange]
}

type Bird implements Animal {
   breed: String
   sightings: [String]
}
```

In the example above, `breed: String` is defined on the interface `Animal` which is implemented by three types. When inspecting `breed` on `Dog` in GraphiQL, there is no documentation because the developer has only documented the `Animal`. To see the comment, the end user must know that the `Animal` interface is documented, and navigate to the type to see the comment. We could save the developer and end-user some time by offering the ability to inherit the documentation comments, when there are no explicit comments on an implementing type. 

What do folks think? Does it make sense that when `Cat.breed` is not documented, to inherit the comments from `Animal.breed`? Is this functionality desired in graphql-java? In this case, since `Cat.breed` has no comments, we could inherit the comment from `Animal.breed`.

**Comment by bbakerman** [comment]
> What do folks think? 

We would be guided by the graphql spec and it says nothing about description inheritance.

So if you gave this schema to some other graphql engine - you could have no expectation that it did description inheritance.

Also there can be multiple interfaces on an object type

```
type Dog implements Animal & Pet {
}
```

which description should be used here?

I think you would be better to use a `graphql.schema.SchemaTransformer` and post process your schema and put in extra description text youself.  We have done that at Atlassian (where I work) successfully to document OUATH scopes that are put in place via directives (eg `@scope(name : "read:thing")`) 

It would be possible to do what you are discussing fairly easily and you get to decide how to resolve edge cases like multiple  interfaces.

**Reply by coleturner** [reply]
Thank you for weighing in @bbakerman 

Your recommendations are great. I had opened this thread hoping for something like that and based on your comments it sounds like this problem isn't unique to what we're doing at Netflix. We thought about adding it to [DGS framework](https://netflix.github.io/dgs/), but first wanted to see whether it could be useful to a wider audience in graphql-java as an optional transformer. 

To the example edge case where `type Dog implements Animal & Pet`, I am thinking out loud that it could be configurable and by default assume that the list of interfaces is searched until the first documentation is found. This would mean `Animal.description` would be used if available. What do you think?
